website/docs/docs/building-a-dbt-project/building-models/python-models.md

```diff
@@ -0,0 +1,378 @@
+</WHCode>
+
+**Note:** Incremental models are not yet supported on BigQuery.
+
+## Python-specific functionality
+
+### Defining functions
+
+In addition to defining a `model` function, the Python model can import other functions, or define its own functions. Here's an example, on Snowpark, defining a custom `add_one` function:
+
+<File name='models/my_python_model.py'>
+
+```python
+def add_one(x):
+    return x + 1
+
+def model(dbt, session):
+    dbt.config(materialized="table")
+    temps_df = dbt.ref("temperatures")
+    
+    # warm things up just a little
+    df = temps_df.withColumn("degree_plus_one", add_one(temps_df["degree"]))
+    return df
+```
+
+</File>
+
+### Using PyPI packages
+
+You can also define functions that depend on installed packages. In this example, we use the `holidays` package to determine if a given date fell on a holiday in France. For simplicity and consistency across platforms, the code uses the Pandas API—it would need to be refactored for scale.
+
+<File name='models/my_python_model.py'>
+
+```python
+import holidays
+import pandas
+
+def is_holiday(date_col):
+    # Chez Jaffle
+    french_holidays = holidays.France()
+    is_holiday = (date_col in french_holidays)
+    return is_holiday
+
+def model(dbt, session):
+    dbt.config(
+        materialized = "table",
+        packages = ["holidays"]
+    )
+    
+    orders_df = dbt.ref("stg_orders")
+    
+    # convert to pandas + lowercase all column names
+    # (uppercase by default on Snowflake)
+    df = orders_df.to_pandas()
+    df.columns= df.columns.str.lower()
+    
+    # apply our function
+    df["is_holiday"] = df["order_date"].apply(is_holiday)
+    
+    # return final dataset
+    return df
+```
+
+</File>
+
+#### Configuring packages
+
+We encourage you to explicitly configure required packages and versions, so that dbt can track them in project metadata. This configuration is required for the implementation on some platforms. If you need specific versions of packages, specify them.
+
+<File name='models/my_python_model.py'>
+
+```python
+def model(dbt, session):
+    dbt.config(
+      packages = ["numpy==1.23.1", "scikit-learn"]
+    )
+```
+
+</File>
+
+<File name='models/config.yml'>
+
+```yml
+version: 2
+
+models:
+  - name: my_python_model
+    config:
+      packages:
+        - "numpy==1.23.1"
+        - scikit-learn
+```
+
+</File>
+
+#### UDFs
+
+You can use the `@udf` decorator or `udf` function to define an "anonymous" function, and call it within your `model` function's dataframe transformation. This is a common pattern for applying more complex functions as data frame operations, especially if those functions require inputs from third-party packages.
+- [Snowpark Python: Creating UDFs](https://docs.snowflake.com/en/developer-guide/snowpark/python/creating-udfs.html)
+- ["PySpark functions: udf"](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.udf.html)
+
+:::info ❔ Our questions
```

Comment:
Should we make the question mark a red one? I couldn't see the white/gray one at first and a red one will probably catch the eye more.

Reply:
Probably overkill for this single page, but we can control how admonitions are rendered: https://docusaurus.io/docs/next/markdown-features/admonitions#customizing-rendering-behavior

Reply:
I played around with a few options, and I think I'm going to go with `:::note` + red question mark for now